synapse/federation/sender/__init__.py

```diff
@@ -197,22 +200,34 @@ async def handle_event(event: EventBase) -> None:
                     if not event.internal_metadata.should_proactively_send():
                         return
 
-                    try:
-                        # Get the state from before the event.
-                        # We need to make sure that this is the state from before
-                        # the event and not from after it.
-                        # Otherwise if the last member on a server in a room is
-                        # banned then it won't receive the event because it won't
-                        # be in the room after the ban.
-                        destinations = await self.state.get_hosts_in_room_at_events(
-                            event.room_id, event_ids=event.prev_event_ids()
-                        )
-                    except Exception:
-                        logger.exception(
-                            "Failed to calculate hosts in room for event: %s",
-                            event.event_id,
+                    # We check the external cache for the destinations, which is
+                    # stored per state group.
```

Comment:
Why not have a `self._use_external_cache` check here like when we set the entries?

Reply:
Oh I guess this is because the external cache just quickly returns `None` if not being used?

Reply:
Yeah, exactly, it's kinda like with logging where you sometimes do `if logger.isEnabledFor(...)` to protect from extra work. It's probably clearer to change the `set` path to always have an `ExternalCache` instance and call a `self._external_cache.is_enabled()`, or something

Reply:
🤷 Either way seems OK, I think having a single specific spot for the enabled logic to live makes sense.